src/index.js

```diff
@@ -0,0 +1,743 @@
+   * @param {Number} fileSize File byte size
+   * @return {Promise<object>} Object containing the correlation ID (`correlationId`) and file ID (`fileId`) of the upload
+   */
+  _finaliseUpload(fileId, fileName, chunksCount, fileSize) {
+    return this.api.send('POST', `v7/file_cmds/upload/${fileId}/finalise_api`, {
+      chunksCount, fileName, fileSize,
+      intent: DEFAULT_UPLOAD_INTENT,
+      sha256: this._sha256
+    })
+      .then(response => response.headers['x-api-correlation-id']);
+  }
+
+  /**
+   * Prepares the remote env to upload a new file
+   * @access private
+   * @return {Promise<string>} Unique file identificator
+   */
+  _prepareUpload() {
+    return this.api.send('POST', 'v7/file_cmds/upload/prepare')
+      .then(response => response.file_id);
+  }
+
+  /**
+   * Saves a media asset in Bynder. If media id is specified in the data a new version of the asset will be saved.
+   * Otherwise a new asset will be saved.
+   * @see {@link https://bynder.docs.apiary.io/#reference/upload-assets/4-finalise-a-completely-uploaded-file/save-as-a-new-asset}
+   * @access private
+   * @param {Object} data Asset data
+   * @param {String} data.fileId File ID
+   * @param {String} data.mediaId Media ID
+   * @param {String} data.name Asset name
+   * @return {Promise<object>}
+   */
+  _saveAsset(data) {
+    const { fileId, mediaId } = data;
+    let url = mediaId ? `api/v4/media/${mediaId}/save/` : 'api/v4/media/save/';
+
+    if (fileId) {
+      url += `${fileId}/`;
+    }
+
+    return this.api.send('POST', url, data);
+  }
+
+  /**
+   * Uploads arbirtrarily sized buffer or stream file to provided S3 endpoint in chunks and registers each chunk to Bynder.
+   * Resolves the passed init result and final chunk number.
+   * @async
+   * @access private
+   * @param {Object} file An object containing the id of the desired collection.
+   * @param {String} file.filename The file name of the file to be saved
+   * @param {Buffer|Readable} file.body The file to be uploaded. Can be either buffer or a read stream.
+   * @param {Number} file.length The length of the file to be uploaded
+   * @param {String} fileId Unique file identifier
+   * @param {Number} size File byte size
+   * @return {Promisee<number>} Total number of chunks uploaded to the upload payload
+   */
+  async _uploadFileInChunks(file, fileId, size) {
+    const { body } = file;
+    // Developers can use this to track file upload progress
+    this._chunks = Math.ceil(size / FILE_CHUNK_SIZE);
+    this._chunkNumber = 0;
+
+    // Iterate over the chunks and send them
+    while (this._chunkNumber <= (this._chunks - 1)) {
+      const start = this._chunkNumber * FILE_CHUNK_SIZE;
+      const end = Math.min(start + FILE_CHUNK_SIZE, size);
+      const chunk = body.slice(start, end);
+      const sha256 = create256HexHash(chunk);
+
+      await this.api.send('POST', `v7/file_cmds/upload/${fileId}/chunk/${this._chunkNumber}`, chunk, {
```

Comment:
Is there any specific reason this call is `await`ed, but none of the other `this.api.send()` calls in this file are?

Reply:
Yup. We want to wait until each chunk is received successfully before sending the next one. This creates a short circuit mechanism that would allow us later to pause the upload, or even re-upload a specific chunk that wasn't uploaded previously.

Reply:
Wouldn't we want this for the other requests as well?